src/mlir/generator.cc

```diff
@@ -93,21 +93,81 @@ namespace mlir::verona
   llvm::Error Generator::parseModule(const ::ast::Ast& ast)
   {
     assert(AST::isClass(ast) && "Bad node");
+    assert(AST::getID(ast) == "$module" && "Bad node");
     module = mlir::ModuleOp::create(getLocation(ast));
-    // TODO: Support more than just functions at the module level
+
     auto body = AST::getClassBody(ast);
-    llvm::SmallVector<::ast::WeakAst, 4> funcs;
-    AST::getSubNodes(funcs, body);
-    for (auto f : funcs)
+    llvm::SmallVector<::ast::WeakAst, 4> globals;
+    AST::getSubNodes(globals, body);
+    for (auto g : globals)
     {
-      auto fun = parseFunction(f.lock());
-      if (auto err = fun.takeError())
-        return err;
-      module->push_back(*fun);
+      if (AST::isClass(g))
+      {
+        auto c = parseClass(g.lock());
+        if (auto err = c.takeError())
+          return err;
+        module->push_back(*c);
+      }
+      else if (AST::isFunction(g))
+      {
+        auto f = parseFunction(g.lock());
+        if (auto err = f.takeError())
+          return err;
+        module->push_back(*f);
+      }
     }
     return llvm::Error::success();
   }
 
+  llvm::Expected<mlir::verona::ClassOp>
+  Generator::parseClass(const ::ast::Ast& ast)
+  {
+    assert(AST::isClass(ast) && "Bad node");
+
+    // Add the type first
+    // Declare before building fields to allow for recursive declaration
+    auto name = AST::getID(ast);
+    auto type = ClassType::get(context, name);
+    typeTable.insert(name, type);
+
+    // Field types
+    llvm::SmallVector<::ast::WeakAst, 1> nodes;
+    AST::getClassTypeElements(ast, nodes);
```

Comment:
`getClassTypeElements` is only used in one place, where you then `zip` something based on the same.  Would it be better to build this more directly.  The function `getClassTypeElements` doesn't seem generally useful, and producing the post-zip version seems to have a more sensible semantics to me.

Reply:
Makes sense.

Reply:
Right, the problem is that the zip auto loop makes use of `parseType`, which is not an AST functionality, but a Generator one, so only available here. I don't want to change the `class T` type to something other than `list<ast>` because all others are like that and, in this case, the best I can do is to have a `list<ast, ast>`.

The whole point of the `ast-utils` is to have very simple (almost useless) helper methods, so we can make the AST manipulation explicit. They will invariably all be inlined now, and most likely replaced by simpler methods in a better AST API, so I'm not worrying if they're not super useful right now, as long as they explicitly state the semantics of the AST traversal.

I will, however, change that code a bit, as it does looks confusing. Hopefully it will be enough.